src/ovs/aca_arp_responder.cpp

```diff
@@ -0,0 +1,415 @@
+#include "aca_arp_responder.h"
+#include "aca_log.h"
+#include "aca_ovs_l2_programmer.h"
+#include "aca_ovs_control.h"
+#include <arpa/inet.h>
+
+using namespace std;
+
+namespace aca_arp_responder
+{
+ACA_ARP_Responder::ACA_ARP_Responder(){
+  _init_arp_db();
+  _init_arp_ofp();
+}
+
+ACA_ARP_Responder::~ACA_ARP_Responder(){
+  _deinit_arp_db();
+  _deinit_arp_ofp();
+}
+
+void ACA_ARP_Responder::_init_arp_db(){
+  try{
+    _arp_db = new unordered_map<arp_entry_data,string,arp_hash>;
+  } catch(const bad_alloc &e){
+    return;
+  }
+}
+
+void ACA_ARP_Responder::_deinit_arp_db(){
+  delete _arp_db;
+  _arp_db = nullptr;
+}
+
+void ACA_ARP_Responder::_init_arp_ofp(){
+  int overall_rc = EXIT_SUCCESS;
+  unsigned long not_care_culminative_time;
+
+  aca_ovs_l2_programmer::ACA_OVS_L2_Programmer::get_instance().execute_openflow_command(
+    "add-flow br-tun \"table=0,priority=25,arp,arp_op=1, actions=CONTROLLER\"",
+    not_care_culminative_time, overall_rc);
+
+  return;
+}
+
+void ACA_ARP_Responder::_deinit_arp_ofp(){
+  unsigned long not_care_culminative_time;
+  int overall_rc = EXIT_SUCCESS;
+
+  aca_ovs_l2_programmer::ACA_OVS_L2_Programmer::get_instance().execute_openflow_command(
+          "del-flows br-tun \"arp,arp_op=1\"",
+          not_care_culminative_time, overall_rc);
+  return;  
+}
+
+ACA_ARP_Responder &ACA_ARP_Responder::get_instance(){
+  static ACA_ARP_Responder instance;
+  return instance;
+}
+
+
+int ACA_ARP_Responder::add_arp_entry(arp_config *arp_cfg_in){
+  arp_entry_data stData;
+
+  if(_validate_arp_entry(arp_cfg_in)){
+    ACA_LOG_ERROR("Valiate arp cfg failed! (mac = %s)\n",
+                  arp_cfg_in->mac_address.c_str());
+    return EXIT_FAILURE;
+  }
+
+  ARP_ENTRY_DATA_SET((arp_entry_data *)&stData, arp_cfg_in);
+
+  _standardize_mac_address(arp_cfg_in->mac_address);
+
+  if (!_search_arp_entry(stData).empty()) {
+    ACA_LOG_ERROR("Entry already existed! (ip = %s)\n",
+                  arp_cfg_in->ipv4_address.c_str());
+    return EXIT_FAILURE;
+  }
+
+  _arp_db_mutex.lock();
+  _arp_db->insert(make_pair(stData,arp_cfg_in->mac_address));
```

Comment:
this unordered_map with mutex lock will work with concurrent threads but will perform poorly, please switch to the new concurrent hashtable and look at the below define in aca_vlan_manager for reference: 

  // hashtable <key: tunnel ID, value: vpc_table_entry>
  CTSL::HashMap<uint, vpc_table_entry *> _vpcs_table;